issue/7528 - Create Separate UI for Viewing Refunds

spencerfinnell opened the pull request:
Fixes #7528 

Proposed Changes:
1. Create separate UI for viewing refunds.

<img width="803" alt="Screen Shot 2020-04-24 at 10 39 58 AM" src="https://user-images.githubusercontent.com/491124/80225056-90820a80-8618-11ea-8a82-562bf1f3f3de.png">
<img width="1120" alt="Screen Shot 2020-04-24 at 10 42 36 AM" src="https://user-images.githubusercontent.com/491124/80225059-91b33780-8618-11ea-9317-7fc16a79f62c.png">

pippinsplugins commented:
@spencerfinnell I got a lot of `Parse error: syntax error, unexpected ')'` errors due to trailing commas. I'm on PHP 7.0.15.

I'm pushing a commit up in a moment to fix them.

pippinsplugins commented on includes/admin/payments/class-payments-table.php:
```diff
@@ -528,7 +528,9 @@ public function column_number( $order ) {
 		$view_url = edd_get_admin_url( array(
 			'page' => 'edd-payment-history',
-			'view' => 'view-order-details',
+			'view' => 'order' === $order->type
```
This is breaking for me. All completed orders are using `view-refund-details`.

My orders all habe a `type` of `sale` not `order`.

spencerfinnell replied:
Oops, nice catch. Fixed.

spencerfinnell commented:
This will need to be refreshed after #7599 is merged.

robincornett commented on includes/admin/payments/refunds.php:
```diff
@@ -0,0 +1,200 @@
+
+		<div class="edit-post-header__settings">
+			<a href="<?php echo esc_url( $order_url ); ?>" class="button button-secondary" autofocus>
+				<?php esc_html_e( 'View Order', 'easy-digital-downloads' ); ?>
```
What would you think about changing this to "View Original Order" instead? (or "Related" but I think original makes more sense in this context).

robincornett commented on includes/admin/payments/refunds.php:
```diff
@@ -0,0 +1,200 @@
+ *
+ * @package     EDD
+ * @subpackage  Admin/Orders
+ * @copyright   Copyright (c) 2018, Easy Digital Downloads, LLC
```
Since this is a new file, should be 2020. Wondering if it should be Sandhills instead of EDD?

robincornett commented on includes/admin/payments/refunds.php:
```diff
@@ -0,0 +1,200 @@
+?>
+
+<div id="edd-order-items" class="postbox edd-edit-purchase-element">
+	<h3 class="hndle">
```
This should be an `h2` to be the correct heading level.

I would remove the class, as `hndle` appears to have a special JS relationship/meaning in WP (for movable boxes, which this is not). Removing it just removes the border underneath, nothing else, so maybe we can implement an edd class if the border is important and there isn't another one we could appropriate.

robincornett commented on includes/admin/payments/refunds.php:
```diff
@@ -0,0 +1,200 @@
+
+<div id="edd-order-items" class="postbox edd-edit-purchase-element">
+	<h3 class="hndle">
+		<span><?php esc_html_e( 'Refund Notes', 'easy-digital-downloads' ); ?></span>
```
Why the `<span>`? Feels unnecessary.

robincornett commented on includes/admin/payments/refunds.php:
```diff
@@ -0,0 +1,200 @@
+?>
+
+<div class="postbox">
+	<h3 class="hndle">
```
See the heading/class/span notes above.

Also not sure about the wording, "Refund Attributes", but not sure I have a better suggestion. Maybe "Details"?

spencerfinnell commented:
@robincornett Addressed most of the feedback, thanks! I'm not sure about 

> this applies to viewing orders as well, but I need to write it here while I'm thinking about it. I do not like that the product titles are linked in the view orders screens (including here). If I'm reviewing an order, I do not necessarily want a primary link to take me off to edit the product.

I'm not convinced one way or the other about if the links should remain, or where they should go, so let's see if there is any other feedback regarding that. Another possibility would be linking them to their relevant Reports page, which may have more of a relation to Sales/Refunds.

robincornett commented on includes/admin/payments/refunds.php:
```diff
@@ -2,11 +2,11 @@
- * @since       3.0
+ * @package    EDD
+ * @subpackage Admin/Orders
+ * @copyright  Copyright (c) 2020, Easy Digital Downloads, LLC
```
I believe this should be Sandhills Development, LLC now.

robincornett commented on includes/admin/payments/refunds.php:
```diff
@@ -166,20 +166,20 @@ function edd_refund_details_notes( $refund ) {
  * @param \EDD\Orders\Order $refund Current Refund.
  */
-function edd_refund_details_attributes( $refund ) {
+function edd_refund_details_details( $refund ) {
```
There are currently not many details here, which is fine if none are needed.

However, one I would request is a line item referencing the original order **number** (and possibly date). I know the button on the screen takes me to the original order, but in my install, the refund order numbers are quite different from the original order numbers.

It would also be nice if the details could show whether this represents a full or partial refund of the original order.

spencerfinnell replied:
> It would also be nice if the details could show whether this represents a full or partial refund of the original order.

Maybe this Refund screen would be good place to have the "Related Refunds" box that was originally shown on the original Order.

LisaCee commented:
> this applies to viewing orders as well, but I need to write it here while I'm thinking about it. I do not like that the product titles are linked in the view orders screens (including here). If I'm reviewing an order, I do not necessarily want a primary link to take me off to edit the product.

I like Robin's idea of having it link to the storefront view.  This is what I expect when viewing a refund elsewhere, like on Amazon.  I can't think of a case where the store owner would want to be linked to the edit page.

robincornett commented:
I can see how a link to the specific download report might be handy--make it easy for a shop owner to see if there is a trend with a certain download.

I'm not sure if I care if the product name in these tables is linked at all--I think potentially adding an action link (or a few) might make more sense, but I know I've inadvertently ended up on the download edit screen more than once from an order screen, and I am not sure it's ever been what I wanted to do. But I'm very willing to concede this could be just me!

LisaCee commented:
Would it be helpful to have the refund id / link be with the line item, maybe as a hover detail?  In this example I have 2 separate refunds in one order and cannot easily tell which refund is which (except for price in this example, but just go with me here).

![image](https://user-images.githubusercontent.com/18077352/80403613-bad00400-8874-11ea-9934-35534d373c38.png)

spencerfinnell commented:
I like the idea of action links under the Download name (especially ensuring the Refund is clearly linked to the relevant line item), but we just need to be careful it's not overloaded. There is already:

Tax: $1.00 | Discount: $1.00 | Copy Download Links

Which could get large: 

Tax: $1.00 | Discount: $1.00 | Copy Download Links | View Download | View Refund

robincornett commented on includes/admin/payments/refunds.php:
```diff
@@ -192,21 +202,33 @@ function edd_refund_details_attributes( $refund ) {
-		<time datetime="<?php echo esc_attr( EDD()->utils->date( $refund->date_created, null, true )->toDateTimeString() ); ?>">
-			<?php echo edd_date_i18n( $refund->date_created, 'M. d, Y' ) . '<br />' . edd_date_i18n( strtotime( $refund->date_created ), 'H:i' ); ?> <?php echo esc_html( edd_get_timezone_abbr() ); ?>
-		</time>
+		<div class="edd-form-group">
```
I feel like my notifications pulled up a different version of the changes and I was seeing the order of the Created/Original Order output reversed, and then was super confused when that was not what I was seeing in the admin. My preference would be to keep the refund creation date/time at the top of the box, and then have the original order number/link underneath (partly because with the "Created" not saying "Refund Created" it would be possible to confuse that data with the original order since that's immediately above it).

robincornett commented on includes/admin/payments/refunds.php:
```diff
@@ -192,21 +202,33 @@ function edd_refund_details_attributes( $refund ) {
+				<?php esc_html_e( 'Created', 'easy-digital-downloads' ); ?>
+			</span>
+			<div class="edd-form-group__control">
+				<time datetime="<?php echo esc_attr( EDD()->utils->date( $refund->date_created, null, true )->toDateTimeString() ); ?>" style="line-height: normal">
```
I've noticed this in other places where EDD is displaying the date and/or time, but is it possible for the date/time to use the format set in General > Settings? I've changed mine a couple different ways to experiment but this appears to be hard coded instead of using my preferences.

spencerfinnell replied:
@robincornett I took the same formatting that is used on the main "Orders" table for the "Date" column. There is some discussion around timezone and dates here: https://github.com/easydigitaldownloads/easy-digital-downloads/issues/7016

The actual format could be changed to your site preference, but for the time being I'm using the Orders table as the source of truth for how to output the date and time.

robincornett replied:
Yes, I asked at one point about creating an issue with the tables not honoring the user settings for dates and times...I think we should honor the settings, with regional differences in how dates are displayed, and 24 hour times potentially confusing some (at least me).
